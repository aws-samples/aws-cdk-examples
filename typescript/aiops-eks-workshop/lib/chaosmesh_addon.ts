// lib/chaosmesh_addon.ts
import { Construct } from 'constructs';
import * as blueprints from '@aws-quickstart/eks-blueprints';
import { setPath } from '@aws-quickstart/eks-blueprints/dist/utils/object-utils';

/**
 * User provided options for the Helm Chart
 */
export interface ChaosMeshAddOnProps extends blueprints.HelmAddOnUserProps {
  version?: string,
  ingressEnabled?: boolean,
  chaosmeshServiceType?: string,
  chaosmeshDashboardListenPort?: string,
}

/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps: blueprints.HelmAddOnProps & ChaosMeshAddOnProps = {
  name: "chaos-mesh",
  namespace: "chaos-testing",
  chart: "chaos-mesh",
  version: "2.4.1",
  release: "chaos-mesh",
  repository:  "https://charts.chaos-mesh.org",
  values: {},

  ingressEnabled: true,
  chaosmeshServiceType: "LoadBalancer",
  chaosmeshDashboardListenPort: "80"
};

/**
 * Main class to instantiate the Helm chart
 */
export class ChaosMeshAddOn extends blueprints.HelmAddOn {

  readonly options: ChaosMeshAddOnProps;

  constructor(props?: ChaosMeshAddOnProps) {
    super({...defaultProps, ...props});
    this.options = this.props as ChaosMeshAddOnProps;
  }

  deploy(clusterInfo: blueprints.ClusterInfo): Promise<Construct> {
    let values: blueprints.Values = populateValues(this.options);
    const chart = this.addHelmChart(clusterInfo, values);

    return Promise.resolve(chart);
  }
}

/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions: ChaosMeshAddOnProps): blueprints.Values {
  const values = helmOptions.values ?? {};

  setPath(values, "ingress.enabled",  helmOptions.ingressEnabled);
  setPath(values, "dashboard.service.type",  helmOptions.chaosmeshServiceType);
  setPath(values, "dashboard.env.LISTEN_PORT", helmOptions.chaosmeshDashboardListenPort);
  setPath(values, "mysql.generate_passwords",  true);

  return values;
}
