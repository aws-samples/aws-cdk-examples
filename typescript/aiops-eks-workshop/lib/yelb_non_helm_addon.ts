// lib/yelb_non_helm_addon.ts
import { Construct } from 'constructs';
import * as blueprints from '@aws-quickstart/eks-blueprints';
import { setPath } from '@aws-quickstart/eks-blueprints/dist/utils/object-utils';
import { KubernetesManifest, KubernetesPatch } from "aws-cdk-lib/aws-eks";
import { loadYaml, readYamlDocument } from "../lib/utils/yaml-utils";

export class YelbNonHelmAddOn implements blueprints.ClusterAddOn {

   deploy(clusterInfo: blueprints.ClusterInfo): void {
        const cluster = clusterInfo.cluster;
        
        const namespace = {
            "apiVersion": "v1",
            "kind": "Namespace",
            "metadata": {
                "name": "yelb"
            }
        }

        const db_deployment = {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-db"
            },
            "spec": {
                "replicas": 1,
                "selector": {
                    "matchLabels": {
                        "app": "yelb-db",
                        "tier": "backenddb"
                    }
                },
                "template": {
                    "metadata": {
                        "labels": {
                            "app": "yelb-db",
                            "tier": "backenddb"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": "yelb-db",
                                "image": "mreferre/yelb-db:0.5",
                                "ports": [
                                    {
                                        "containerPort": 5432
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        }
        
        const db_service = {
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-db",
                "labels": {
                    "app": "yelb-db",
                    "tier": "backenddb"
                }
            },
            "spec": {
                "type": "ClusterIP",
                "ports": [
                    {
                        "port": 5432
                    }
                ],
                "selector": {
                    "app": "yelb-db",
                    "tier": "backenddb"
                }
            }
        }
        
        const redis_deployment = {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "namespace": "yelb",
                "name": "redis-server"
            },
            "spec": {
                "selector": {
                    "matchLabels": {
                        "app": "redis-server",
                        "tier": "cache"
                    }
                },
                "replicas": 1,
                "template": {
                    "metadata": {
                        "labels": {
                            "app": "redis-server",
                            "tier": "cache"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": "redis-server",
                                "image": "redis:4.0.2",
                                "ports": [
                                    {
                                        "containerPort": 6379
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        }
        const redis_service = {
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "namespace": "yelb",
                "name": "redis-server",
                "labels": {
                    "app": "redis-server",
                    "tier": "cache"
                }
            },
            "spec": {
                "type": "ClusterIP",
                "ports": [
                    {
                        "port": 6379
                    }
                ],
                "selector": {
                    "app": "redis-server",
                    "tier": "cache"
                }
            }
        }
        
        const app_deployment = {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-appserver"
            },
            "spec": {
                "replicas": 1,
                "selector": {
                    "matchLabels": {
                        "app": "yelb-appserver",
                        "tier": "middletier"
                    }
                },
                "template": {
                    "metadata": {
                        "labels": {
                            "app": "yelb-appserver",
                            "tier": "middletier"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": "yelb-appserver",
                                "image": "mreferre/yelb-appserver:0.5",
                                "ports": [
                                    {
                                        "containerPort": 4567
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        }
        const app_service = {
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-appserver",
                "labels": {
                    "app": "yelb-appserver",
                    "tier": "middletier"
                }
            },
            "spec": {
                "type": "ClusterIP",
                "ports": [
                    {
                        "port": 4567
                    }
                ],
                "selector": {
                    "app": "yelb-appserver",
                    "tier": "middletier"
                }
            }
        }
        
        const ui_deployment = {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-ui"
            },
            "spec": {
                "replicas": 1,
                "selector": {
                    "matchLabels": {
                        "app": "yelb-ui",
                        "tier": "frontend"
                    }
                },
                "template": {
                    "metadata": {
                        "labels": {
                            "app": "yelb-ui",
                            "tier": "frontend"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": "yelb-ui",
                                "image": "mreferre/yelb-ui:0.7",
                                "ports": [
                                    {
                                        "containerPort": 80
                                    }
                                ],
                                "resources": {
                                    "limits": {
                                        "cpu": "50m",
                                        "memory": "32Mi"
                                    },
                                    "requests": {
                                        "cpu": "25m",
                                        "memory": "16Mi"
                                    }
                                }
                            }
                        ]
                    }
                }
            }
        }
        const ui_service = {
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "namespace": "yelb",
                "name": "yelb-ui",
                "labels": {
                    "app": "yelb-ui",
                    "tier": "frontend"
                }
            },
            "spec": {
                "type": "LoadBalancer",
                "ports": [
                    {
                        "port": 80,
                        "protocol": "TCP",
                        "targetPort": 80
                    }
                ],
                "selector": {
                    "app": "yelb-ui",
                    "tier": "frontend"
                }
            }
        }

        const db_latency = {
            "apiVersion": "chaos-mesh.org/v1alpha1",
            "kind": "NetworkChaos",
            "metadata": {
                "name": "latency-db"
            },
            "spec": {
                "action": "delay",
                "delay": {
                    "correlation": "0",
                    "jitter": "0ms",
                    "latency": "300ms"
                },
                "direction": "to",
                "duration": "1h",
                "mode": "all",
                "selector": {
                    "labelSelectors": {
                        "tier": "backenddb"
                    },
                    "namespaces": [
                        "yelb"
                    ]
                }
            }
        }
        
        const ui_burn_cpu = {
          "apiVersion": "chaos-mesh.org/v1alpha1",
          "kind": "StressChaos",
          "metadata": {
            "name": "ui-burn-cpu"
          },
          "spec": {
            "mode": "all",
            "selector": {
              "namespaces": [
                "yelb"
              ],
              "labelSelectors": {
                "tier": "frontend"
              }
            },
            "stressors": {
              "cpu": {
                "workers": 2,
                "load": 100,
                "options": [
                  "--cpu 2",
                  "--hdd 1"
                ]
              }
            },
            "duration": "6h"
          }
        }
        
        new KubernetesManifest(cluster.stack, "yelb-manifest", {
            cluster,
            manifest: [
                namespace,
                db_deployment,
                db_service,
                redis_deployment, 
                redis_service, 
                app_deployment,
                app_service,
                ui_deployment, 
                ui_service,
                
                // ChaosMesh fault injections
                db_latency,
                ui_burn_cpu
                ],
            overwrite: true
        });
        
    }
}