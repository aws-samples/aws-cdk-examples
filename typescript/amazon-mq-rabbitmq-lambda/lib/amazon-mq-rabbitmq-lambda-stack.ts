import { Duration, Stack, StackProps, CfnOutput, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Function, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { InstanceClass, InstanceSize, InstanceType } from 'aws-cdk-lib/aws-ec2';
import { RabbitMqBrokerEngineVersion, RabbitMqBrokerInstance, RabbitMqEventSource } from '@cdklabs/cdk-amazonmq';

export class AmazonMqRabbitmqLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Define the admin secret in AWS Secrets Manager
    const adminSecret = new secretsmanager.Secret(this, 'AdminSecret', {
      secretName: 'AdminCredentials',
      generateSecretString: {
        secretStringTemplate: JSON.stringify({
          username: 'admin' // Set a default username for RabbitMQ broker
        }),
        generateStringKey: 'password', // Auto-generate password
        excludePunctuation: true,      // Avoid punctuation in password
        passwordLength: 12,            // Set password length to 12 characters
      },
      removalPolicy: RemovalPolicy.DESTROY, // Ensure the secret is deleted on stack destroy
    });

    // Create a RabbitMQ broker instance with specified version and instance type
    const broker = new RabbitMqBrokerInstance(this, 'RabbitMqBroker', {
      publiclyAccessible: true,                 // Publicly accessible RabbitMQ broker
      version: RabbitMqBrokerEngineVersion.V3_13, // Use RabbitMQ version 3.13
      instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO), // Instance type T3.micro
      admin: {
        username: adminSecret.secretValueFromJson('username').unsafeUnwrap(), // Use username from Secrets Manager
        password: adminSecret.secretValueFromJson('password'),               // Use password from Secrets Manager
      },
      autoMinorVersionUpgrade: true, // Enable auto minor version upgrades
    });

    // Output the AMQP and Web Console endpoints and ports for the RabbitMQ broker
    new CfnOutput(this, 'AmqpEndpointUrl', { value: broker.endpoints.amqp.url });
    new CfnOutput(this, 'AmqpEndpointPort', { value: broker.endpoints.amqp.port.toString() });
    new CfnOutput(this, 'WebConsoleUrl', { value: broker.endpoints.console.url });
    new CfnOutput(this, 'WebConsolePort', { value: broker.endpoints.console.port.toString() });

    // Create a custom CloudWatch Log Group for the consumer Lambda function
    const consumerLambdaLogGroup = new logs.LogGroup(this, 'ConsumerLambdaLogGroup', {
      logGroupName: 'customLogGroup', // Custom log group name
      removalPolicy: RemovalPolicy.DESTROY, // Ensure it's deleted on stack destroy
    });

    // Define the consumer Lambda function which will handle messages from RabbitMQ
    const consumer_lambda = new Function(this, 'consumer_lambdaFunction', {
      runtime: Runtime.NODEJS_20_X,       // Use Node.js 20.x runtime for the Lambda function
      code: Code.fromAsset('lambda'),     // Path to Lambda function code directory
      handler: 'consumer.handler',        // The entry point (handler) for the Lambda function
      memorySize: 128,                    // Set memory size to 128 MB
      timeout: Duration.seconds(30),      // Set timeout duration to 30 seconds
      logGroup: consumerLambdaLogGroup,   // Attach the custom CloudWatch log group
    });

    consumer_lambda.logGroup.applyRemovalPolicy(RemovalPolicy.DESTROY);

    // Add RabbitMQ as an event source for the Lambda function to consume messages from 'testQueue'
    consumer_lambda.addEventSource(new RabbitMqEventSource({
      broker,                           // Reference to the RabbitMQ broker instance
      credentials: adminSecret,         // Use admin credentials from Secrets Manager
      queueName: 'testQueue',           // Queue name in RabbitMQ from which Lambda will consume messages
    }));
  }
}
