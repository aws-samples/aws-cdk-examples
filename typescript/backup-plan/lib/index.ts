import {
  Duration,
  CfnOutput,
  aws_backup as bk,
  aws_events as events,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface BackupProps {
  /**
   * Resources to apply backup plan.
   */
  readonly resources: bk.BackupResource[];

  /**
   * The display name of the backup plan.
   */
  readonly backupPlanName: string;

  /**
   * The duration after a backup job is successfully
   * started before it must be completed or it is
   * canceled by AWS Backup.
   *
   * Note: `backupCompletionWindow` must be at least 60 minutes greater
   * than @backupStartWindows
   *
   * @default - 3 hours
   */
  readonly backupCompletionWindow?: Duration;

  /**
   * The duration after a backup is scheduled before
   * a job is canceled if it doesn't start successfully.
   *
   * @default - 1 hour less than @backupCompletionWindow
   */
  readonly backupStartWindow?: Duration;

  /**
   * Specifies the duration after creation that a recovery point is deleted.
   * Must be greater than moveToColdStorageAfter.
   *
   * @default - 30 days
   */
  readonly deleteBackupAfter?: Duration;

  /**
   * Specifies the duration after creation that a recovery point is moved to cold storage.
   *
   * @default - recovery point is never moved to cold storage
   */
  readonly moveBackupToColdStorageAfter?: Duration;

  /**
   * How frequently backup jobs would be started.
   *
   * @default - 24 hours
   */
  readonly backupRateHour?: number;
}

/**
 * Construct to create a Backup Plan with specific backing cadence.
 *
 * @stability stable
 */
export class Backup extends Construct {
  /**
   * Backup plan
   */
  public readonly backupPlan: bk.BackupPlan;
  /**
   * 
   * @param scope Construct's scope
   * @param id Construct's id
   * @param props Construct's props
   */
  constructor(scope: Construct, id: string, props: BackupProps) {
    super(scope, id);

    const hourlyRate = `0/${props.backupRateHour || 24}`;

    const completionWindow = props.backupCompletionWindow || Duration.hours(3);
    const startWindow = props.backupStartWindow || Duration.hours(completionWindow.toHours() - 1);

    if ((completionWindow.toHours() - startWindow.toHours()) < 1)
      throw Error('Backup completion window must be at least 60 minutes greater than backup start window');

    const scheduledBkRule = new bk.BackupPlanRule({
      completionWindow,
      startWindow,
      deleteAfter: props.deleteBackupAfter || Duration.days(30),
      // Only cron expressions are supported
      scheduleExpression: events.Schedule.cron({
        minute: '0',
        hour: hourlyRate,
      }),
      moveToColdStorageAfter: props.moveBackupToColdStorageAfter,
    });

    this.backupPlan = new bk.BackupPlan(this, 'BackupPlan', {
      backupPlanName: props.backupPlanName,
      backupPlanRules: [scheduledBkRule],
    });

    this.backupPlan.addSelection('BackupSelection', {
      resources: props.resources,
    });

    // Outputs
    const outputVars = {
      BackupPlanId: this.backupPlan.backupPlanId,
      BackupPlanArn: this.backupPlan.backupPlanArn,
    };
    Object.entries(outputVars).forEach(
      ([outName, outValue]) => new CfnOutput(this, outName, { value: outValue }),
    );
  }
}
