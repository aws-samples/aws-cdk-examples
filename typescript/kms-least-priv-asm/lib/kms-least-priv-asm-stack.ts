import * as cdk from 'aws-cdk-lib';
import { AccountPrincipal, ArnPrincipal, CompositePrincipal, Effect, ManagedPolicy, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import { Function, InlineCode, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import * as fs from "fs";

export class KmsLeastPrivAsmStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // -- 1. Create a KMS for Secrets Manager -- //

    const kmsKey = new Key(this, "SecretsManagerKMS", {
      alias: "kms/app1/secretsmanager" + this.stackName,
      description: "KMS to manage App1 keys in Secrets Manager",
    });
    const SecretsManagerPolicy = new PolicyStatement();
    SecretsManagerPolicy.addAnyPrincipal();
    SecretsManagerPolicy.addActions(
        'kms:Encrypt',
        'kms:Decrypt',
        'kms:ReEncrypt*',  
        'kms:CreateGrant',
        'kms:DescribeKey'
    );
    SecretsManagerPolicy.addResources('*');
    SecretsManagerPolicy.addCondition("StringEquals", {"kms:CallerAccount" : this.account});
    SecretsManagerPolicy.addCondition("StringEquals", {"kms:ViaService" : "secretsmanager." + this.region + ".amazonaws.com"});

    kmsKey.addToResourcePolicy(SecretsManagerPolicy);

    // -- 2. Create a randomly generated secret -- //

    const asmSecret = new Secret(this, "App1Secret", {
      secretName: "secret/for/lambda" + this.stackName,
      encryptionKey: kmsKey,
      generateSecretString: {
        secretStringTemplate: JSON.stringify({ database: "testDatabase", username: "testAppUser" }),
        generateStringKey: "password",
        excludeCharacters: '\"@/\\',
        passwordLength: 22
      },
    });

    // -- 3. Create a SSM parameter with ASM arn -- //

    const ssmParam = new StringParameter(this, "StringParameter", {
      description: "Some user-friendly description for App1 secrets",
      parameterName: "/path/to/asm/secrets/app1" + this.stackName,
      simpleName: false,
      stringValue: asmSecret.secretArn,
    });
    

    // -- 4. Create an IAM role used to fetch secret value via SSM parameter -- //

    const lambdaExeRole = new Role(this, "LambdaExecutionRole", {
      assumedBy: new CompositePrincipal(
        new ServicePrincipal("lambda.amazonaws.com"),
        new AccountPrincipal(this.account)
      ),      
    });

    // Then add the various resources we'll need

    lambdaExeRole.addManagedPolicy(ManagedPolicy.fromManagedPolicyArn(this, "AWSLambdaBasicExecutionRole", "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"));
    asmSecret.grantRead(lambdaExeRole);
    ssmParam.grantRead(lambdaExeRole);

    // Updating ASM resource policy to allow access to the secret

    asmSecret.addToResourcePolicy(
      new PolicyStatement({
        actions: [
          'secretsmanager:GetSecretValue',
          'secretsmanager:DescribeSecret',
          'secretsmanager:ListSecretVersionIds'
        ],
        effect: Effect.ALLOW,
          principals: [new ArnPrincipal(lambdaExeRole.roleArn)],
        resources: ['*']
      })
    );

    // -- 5. Create a lambda to fetch secret and use SSM parameter -- //

    const asmFetchLambda = new Function(this, "asmFetchLambda", {
      functionName: "asmFetchLambda" + this.stackName,
      description: "Sample lambda to check if an IAM role can fetch ASM secret",
      runtime: Runtime.PYTHON_3_9,
      handler: "index.lambda_handler",
      environment: {
          "SSMParaName": ssmParam.parameterName
      },
      code: new InlineCode(fs.readFileSync("./lambda/lambda_function.py", { encoding: "utf-8" })),
      role: lambdaExeRole,
    });

    new cdk.CfnOutput(this, "Sample Lambda Name", {
      value: asmFetchLambda.functionName,
        description: "The name of the Lambda function",
      exportName: "sampleLambdaName",
    });

  }
}
