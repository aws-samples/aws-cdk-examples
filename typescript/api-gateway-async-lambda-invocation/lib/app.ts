#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { ApiGatewayAsyncLambdaStack } from '../lib/api-gateway-async-lambda-invocation-stack';

const app = new cdk.App();
const env = { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION }
const prefix = 'apigw-async-lambda';

const apigw_async_lambda = new ApiGatewayAsyncLambdaStack(app, 'ApiGatewayAsyncLambdaStack', {
  env,
  stackName: `${prefix}-stack`,
  prefix,
});
