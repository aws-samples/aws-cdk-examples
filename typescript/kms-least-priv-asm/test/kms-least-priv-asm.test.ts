import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as KmsLeastPrivAsm from '../lib/kms-least-priv-asm-stack';

test('KMS key and ASM Test', () => {
  const app = new cdk.App();
    // WHEN
  const stack = new KmsLeastPrivAsm.KmsLeastPrivAsmStack(app, 'MyTestStack');
    // THEN
  const template = Template.fromStack(stack);
    // Assert the template matches the snapshot.
  expect(template.toJSON()).toMatchSnapshot();
});
