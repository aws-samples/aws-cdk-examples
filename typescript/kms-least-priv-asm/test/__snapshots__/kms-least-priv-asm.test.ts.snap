// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KMS key and ASM Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "App1Secret46548AA8": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "GenerateSecretString": {
          "ExcludeCharacters": ""@/\\",
          "GenerateStringKey": "password",
          "PasswordLength": 22,
          "SecretStringTemplate": "{"database":"testDatabase","username":"testAppUser"}",
        },
        "KmsKeyId": {
          "Fn::GetAtt": [
            "SecretsManagerKMSBF4037C3",
            "Arn",
          ],
        },
        "Name": "secret/for/lambdaMyTestStack",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaExecutionRoleD5C26073": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaExecutionRoleDefaultPolicy6D69732F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "App1Secret46548AA8",
              },
            },
            {
              "Action": [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter",
                    {
                      "Ref": "StringParameter472EED0E",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaExecutionRoleDefaultPolicy6D69732F",
        "Roles": [
          {
            "Ref": "LambdaExecutionRoleD5C26073",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SecretsManagerKMSAlias75005233": {
      "Properties": {
        "AliasName": "alias/kms/app1/secretsmanagerMyTestStack",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "SecretsManagerKMSBF4037C3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "SecretsManagerKMSBF4037C3": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "KMS to manage App1 keys in Secrets Manager",
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:ReEncrypt*",
                "kms:CreateGrant",
                "kms:DescribeKey",
              ],
              "Condition": {
                "StringEquals": {
                  "kms:CallerAccount": {
                    "Ref": "AWS::AccountId",
                  },
                  "kms:ViaService": {
                    "Fn::Join": [
                      "",
                      [
                        "secretsmanager.",
                        {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Condition": {
                "StringEquals": {
                  "kms:ViaService": {
                    "Fn::Join": [
                      "",
                      [
                        "secretsmanager.",
                        {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:CreateGrant",
                "kms:DescribeKey",
              ],
              "Condition": {
                "StringEquals": {
                  "kms:ViaService": {
                    "Fn::Join": [
                      "",
                      [
                        "secretsmanager.",
                        {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": "kms:Decrypt",
              "Condition": {
                "StringEquals": {
                  "kms:ViaService": {
                    "Fn::Join": [
                      "",
                      [
                        "secretsmanager.",
                        {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "LambdaExecutionRoleD5C26073",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "StringParameter472EED0E": {
      "Properties": {
        "Description": "Some user-friendly description for App1 secrets",
        "Name": "/path/to/asm/secrets/app1MyTestStack",
        "Type": "String",
        "Value": {
          "Ref": "App1Secret46548AA8",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "asmFetchLambda3716BFBE": {
      "DependsOn": [
        "LambdaExecutionRoleDefaultPolicy6D69732F",
        "LambdaExecutionRoleD5C26073",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "import logging
import boto3
import base64
from botocore.exceptions import ClientError
import os

logger = logging.getLogger(__name__)


def lambda_handler(event, context):

     
    ssm = boto3.client('ssm')
    parameterName = os.environ['SSMParaName']
    parameterValue = ssm.get_parameter(Name=parameterName)
    
    result = get_secret(parameterValue['Parameter']['Value'])
    return result
        
def get_secret(secretId):

    session = boto3.session.Session()
    client = session.client('secretsmanager')

    # In this sample we only handle the specific exceptions for the 'GetSecretValue' API.
    # See https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html
    # We rethrow the exception by default.

    try:
        get_secret_value_response = client.get_secret_value(
            SecretId=secretId
        )
    except ClientError as e:
        logger.error(e)
        if e.response['Error']['Code'] == 'DecryptionFailureException':
            # Secrets Manager can't decrypt the protected secret text using the provided KMS key.
            # Deal with the exception here, and/or rethrow at your discretion.
            raise e
        elif e.response['Error']['Code'] == 'InternalServiceErrorException':
            # An error occurred on the server side.
            # Deal with the exception here, and/or rethrow at your discretion.
            raise e
        elif e.response['Error']['Code'] == 'InvalidParameterException':
            # You provided an invalid value for a parameter.
            # Deal with the exception here, and/or rethrow at your discretion.
            raise e
        elif e.response['Error']['Code'] == 'InvalidRequestException':
            # You provided a parameter value that is not valid for the current state of the resource.
            # Deal with the exception here, and/or rethrow at your discretion.
            raise e
        elif e.response['Error']['Code'] == 'ResourceNotFoundException':
            # We can't find the resource that you asked for.
            # Deal with the exception here, and/or rethrow at your discretion.
            raise e
    else:
        # Decrypts secret using the associated KMS CMK.
        # Depending on whether the secret is a string or binary, one of these fields will be populated.
        if 'SecretString' in get_secret_value_response:
            return get_secret_value_response['SecretString']
        else:
            return base64.b64decode(get_secret_value_response['SecretBinary'])
",
        },
        "Description": "Sample lambda to check if an IAM role can fetch ASM secret",
        "Environment": {
          "Variables": {
            "SSMParaName": {
              "Ref": "StringParameter472EED0E",
            },
          },
        },
        "FunctionName": "asmFetchLambdaMyTestStack",
        "Handler": "index.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRoleD5C26073",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
